katdal/datasources.py

```diff
@@ -263,6 +263,36 @@ def _ensure_prefix_is_set(chunk_info, telstate):
     return chunk_info
 
 
+def _upgrade_chunk_info(chunk_info, improved_chunk_info):
+    """Replace chunk info items with better ones while preserving number of dumps."""
+    for key, improved_info in improved_chunk_info.items():
+        original_info = chunk_info.get(key, improved_info)
+        n_original_dumps = original_info['shape'][0]
+        chunks = improved_info['chunks']
+        time_chunks = chunks[0]
+        n_improved_dumps = sum(time_chunks)
+        if n_improved_dumps != n_original_dumps:
+            logger.debug("Original '%s' array has %d dumps while improved "
+                         "version has %d - forcing it to %d dumps", key,
+                         n_original_dumps, n_improved_dumps, n_original_dumps)
+            # Extend/truncate improved version to match original number of dumps
+            if n_improved_dumps < n_original_dumps:
+                time_chunks += (n_original_dumps - n_improved_dumps) * (1,)
+            else:
+                while time_chunks and sum(time_chunks) > n_original_dumps:
+                    time_chunks = time_chunks[:-1]
```

Comment:
This algorithm is quadratic time (both because you're summing each time, and because you're creating a new tuple each time). You can make it linear by iterating through the chunks and keeping a running total until the running total is >= n_original_dumps.

Reply:
Good point - I was being lazy.